File: captainhcg/go/174-dungeon-game.go
func min(a, b int) int{
    if a < b{
        return a
    }
    return b
}

func max(a, b int) int{
    if a > b{
        return a
    }
    return b
}

func calculateMinimumHP(dungeon [][]int) int {
	mat := make([][]int, len(dungeon))
	for idx := range dungeon {
		mat[idx] = make([]int, len(dungeon[0]))
	}
	for row := len(dungeon) - 1; row >= 0; row -= 1 {
		for col := len(dungeon[0]) - 1; col >= 0; col -= 1 {
			var minhp int
			if row == len(dungeon)-1 && col == len(dungeon[0])-1 {
				minhp = -dungeon[row][col]
			} else if row == len(dungeon)-1 {
				minhp = -dungeon[row][col] + mat[row][col+1]
			} else if col == len(dungeon[0])-1 {
				minhp = -dungeon[row][col] + mat[row+1][col]
			} else {
				minhp = -dungeon[row][col] + min(mat[row+1][col], mat[row][col+1])
			}
			mat[row][col] = max(minhp, 0)
		}
	}
	return mat[0][0] + 1
}

File: bwu/dp/174-dungeon-game.py
class Solution(object):
    def calculateMinimumHP(self, dungeon):
        """
        :type dungeon: List[List[int]]
        :rtype: int
        """
        if not dungeon or not dungeon[0]:
            return 0
            
        m, n = len(dungeon), len(dungeon[0])
        dp = [[0 for _ in xrange(n)] for _ in xrange(m)]
        
        dp[m - 1][n - 1] = 1 if dungeon[m - 1][n - 1] >= 0 else 1 - dungeon[m - 1][n - 1]
        for i in xrange(m - 2, -1, -1):
            dp[i][n - 1] = max(1, dp[i + 1][n - 1] - dungeon[i][n - 1])
        for j in xrange(n - 2, -1, -1):
            dp[m - 1][j] = max(1, dp[m - 1][j + 1] - dungeon[m - 1][j])
        for i in xrange(m - 2, -1, -1):
            for j in xrange(n - 2, -1, -1):
                dp[i][j] = max(1, min(dp[i + 1][j], dp[i][j + 1]) - dungeon[i][j])
        return dp[0][0]
